packages/app/src/elements/audit.ts

```diff
@@ -0,0 +1,214 @@
+                                <pl-vault-item-list-item
+                                    .item=${listItem.item}
+                                    .vault=${listItem.vault}
+                                    class="stretch collapse"
+                                ></pl-vault-item-list-item>
+                            </div>
+                        `
+                    )}
+                </pl-list>
+                <pl-button
+                    class="slim margined transparent"
+                    @click=${() => this.go("items", routeParams)}
+                    ?hidden=${listItems.length < 6}
+                >
+                    <div>${$l("Show All")}</div>
+                    <pl-icon icon="arrow-right" class="small left-margined"></pl-icon>
+                </pl-button>
+            </section>
+        `;
+    }
+}
+
+@customElement("pl-audit")
+export class Audit extends StateMixin(LitElement) {
+    @state()
+    private _reusedPasswords: ListItem[] = [];
+
+    @state()
+    private _weakPasswords: ListItem[] = [];
+
+    @state()
+    private _compromisedPasswords: ListItem[] = [];
+
+    @queryAll("pl-audit-list-item")
+    private _countElements: HTMLDivElement[];
+
+    static styles = [
+        shared,
+        css`
+            :host {
+                display: block;
+            }
+
+            .counts {
+                display: grid;
+                grid-gap: 1em;
+                margin: 1em;
+                grid-template-columns: repeat(auto-fit, minmax(20em, 1fr));
+            }
+
+            @media (max-width: 700px) {
+                .counts {
+                    grid-template-columns: repeat(auto-fit, minmax(15em, 1fr));
+                }
+            }
+        `,
+    ];
+
+    render() {
+        const { _reusedPasswords, _weakPasswords, _compromisedPasswords } = this;
+
+        // TODO: Remove this
+        console.log("======== rendering!");
+        console.log(
+            JSON.stringify({
+                reusedPasswordsCount: _reusedPasswords.length,
+                weakPasswordsCount: _weakPasswords.length,
+                compromisedPasswordsCount: _compromisedPasswords.length,
+            })
+        );
+
+        return html`
+            <div class="padded layout">
+                <div class="vertical spacing">
+                    <h3>
+                        Here you can see some counts for passwords you should change, as they might be easier to be
+                        easily guessed.
+                    </h3>
+                    <p class="subtle">Click a count to view the affected items.</p>
+                    <div class="counts">
+                        <pl-audit-list-item
+                            .listItems=${_reusedPasswords}
+                            .label=${$l("Reused Passwords")}
+                            .routeParams=${{ auditReused: "true" }}
+                        ></pl-audit-list-item>
```

Comment:
Again, no reason to create a separate element here. If you want to avoid repetition, you can use a partial render function, like so:

```ts
private _renderSection(listItems: ListItem[], label: string) {
    return html`
            <section class="box count" ?hidden=${listItems.length === 0}>
                <h2 class="uppercase bg-dark border-bottom semibold center-aligning spacing horizontal layout">
                    <div></div>
                    <div>${label}</div>
                    <div class="tiny bold tag">${listItems.length}</div>
                    <div class="stretch"></div>
                </h2>
                <pl-list>
                    ${listItems.slice(0, 5).map(
                        (listItem) => html`
                            <div class="list-item hover click" @click=${() => this.go(`items/${listItem.item.id}`)}>
                                <pl-vault-item-list-item
                                    .item=${listItem.item}
                                    .vault=${listItem.vault}
                                    class="stretch collapse"
                                ></pl-vault-item-list-item>
                            </div>
                        `
                    )}
                </pl-list>
                <pl-button
                    class="slim margined transparent"
                    @click=${() => this.go("items", routeParams)}
                    ?hidden=${listItems.length < 6}
                >
                    <div>${$l("Show All")}</div>
                    <pl-icon icon="arrow-right" class="small left-margined"></pl-icon>
                </pl-button>
            </section>
    `;
}

render() {
    return html`
        ...
        ${this._renderSection(this._reusedPasswords, $l("Reused Passwords")}
        ${this._renderSection(this._weakPassword, $l("Weak Passwords")}
        ${this._renderSection(this._compromisedPasswords, $l("Compromised Passwords")}
        ...
    `;
}
```

But even that might be "premature generalisation". I found that often it's better to repeat yourself a little than unnecessarily complicating things by trying to deduplicate code.

Reply:
Gotcha, I'm happy with the functions stuff. Refactoring this piece actually helped me a lot with tweaking the design, so I wouldn't call it premature :blush: